nixos/modules/services/networking/netmaker/netmaker.nix

```
@@ -0,0 +1,787 @@
+      defaultText = lib.literalExpression ''"''${config.services.netmaker.secretsDir}/MQ_PASSWORD"'';
+      description = lib.mdDoc ''
+        Location of the file holding password for the message queue (auto-generated by init script).
+      '';
+    };
+    mqtt.acl.restricted = lib.mkEnableOption "MQTT instance access restrictions for `netmaker` user";
+    mqtt.acl.extraRules = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = [ ];
+      description = lib.mdDoc ''
+        Extends list of ACL rules for `netmaker`.
+      '';
+      example = lib.literalExpression ''
+        [
+          "readwrite some/missing/topic"
+        ]
+      '';
+    };
+
+    debug = lib.mkEnableOption "configuring everything with the most verbose log level";
+    debugTools = lib.mkOption {
+      type = with lib.types; bool;
+      default = config.services.netmaker.debug;
+      defaultText = lib.literalExpression ''config.services.netmaker.debug'';
+      description = lib.mdDoc ''
+        Install additional debugging tools:
+        - `mqttui`
+        - `netmaker-mqttui` script
+      '';
+    };
+    coredns.debug = lib.mkOption {
+      type = with lib.types; bool;
+      default = config.services.netmaker.debug;
+      defaultText = lib.literalExpression ''config.services.netmaker.debug'';
+      description = lib.mdDoc ''
+        Run CoreDNS in a full debug mode.
+      '';
+    };
+    webserver.debug = lib.mkOption {
+      type = with lib.types; bool;
+      default = config.services.netmaker.debug;
+      defaultText = lib.literalExpression ''config.services.netmaker.debug'';
+      description = lib.mdDoc ''
+        Run webserver in a full debug mode.
+      '';
+    };
+    mqtt.debug = lib.mkOption {
+      type = with lib.types; bool;
+      default = config.services.netmaker.debug;
+      defaultText = lib.literalExpression ''config.services.netmaker.debug'';
+      description = lib.mdDoc ''
+        Run message queue in a full debug mode.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable (lib.mkMerge [
+    {
+      environment.systemPackages = [ cfg.nmctl.package ];
+
+      networking.firewall.allowedUDPPortRanges = with cfg.firewall.ports;  [{
+        from = networks.start;
+        to = networks.start + networks.capacity - 1;
+      }];
+    }
+    {
+      systemd.tmpfiles.rules = [
+        "d ${cfg.secretsDir} 1700 root root"
+        "f ${envFile.generated.config} 1644 root root"
+        "f ${envFile.generated.secrets} 1600 root root"
+        "f ${envFile.user.config} 1644 root root"
+        "f ${envFile.user.secrets} 1600 root root"
+      ];
+
+      services.netmaker.environmentFiles = [
+        envFile.generated.config
+        envFile.generated.secrets
+        envFile.user.config
+        envFile.user.secrets
+        (lib.mkAfter cfg.environment)
+      ];
+
+      services.netmaker.config = {
+        server.apiconn = "${cfg.api.domain}:443";
+        server.apihost = cfg.api.domain;
```

```suggestion
      services.netmaker.config.server = {
        apiconn = "${cfg.api.domain}:443";
        apihost = cfg.api.domain;
```
we can save lots of repetition by moving server

---

yeah, but makes it a lot harder to track down where the non-NixOS configuration starts and how it maps to the Netmaker configuration. How about something like this?

```
services.netmaker.config = {
	server = {
		....
	};
};
```